libraries/AP_Torqeedo/AP_Torqeedo.cpp

```diff
@@ -0,0 +1,310 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "AP_Torqeedo.h"
+
+#if TORQEEDO_ENABLED
+
+#include <AP_Common/AP_Common.h>
+#include <AP_Math/AP_Math.h>
+#include <SRV_Channel/SRV_Channel.h>
+
+#define TORQEEDO_SERIAL_BAUD        19200   // communication is always at 19200
+#define TORQEEDO_PACKET_HEADER      0xAC    // communication packet header
+#define TORQEEDO_PACKET_FOOTER      0xAD    // communication packer footer
+
+extern const AP_HAL::HAL& hal;
+
+// parameters
+const AP_Param::GroupInfo AP_Torqeedo::var_info[] = {
+
+    // @Param: ENABLE
+    // @DisplayName: Torqeedo enable
+    // @Description: Torqeedo enable
+    // @Values: 0:Disabled, 1:Enabled
+    // @User: Standard
+    // @RebootRequired: True
+    AP_GROUPINFO_FLAGS("ENABLE", 1, AP_Torqeedo, _enable, 0, AP_PARAM_FLAG_ENABLE),
+
+    // @Param: ONOFF_PIN
+    // @DisplayName: Torqeedo ON/Off pin
+    // @Description: Pin number connected to Torqeedo's on/off pin. -1 to disable turning motor on/off from autopilot
+    // @Values: -1:Disabled,50:AUX1,51:AUX2,52:AUX3,53:AUX4,54:AUX5,55:AUX6
+    // @User: Standard
+    // @RebootRequired: True
+    AP_GROUPINFO("ONOFF_PIN", 2, AP_Torqeedo, _pin_onoff, -1),
+
+    // @Param: DE_PIN
+    // @DisplayName: Torqeedo DE pin
+    // @Description: Pin number connected to RS485 to Serial converter's DE pin. -1 to disable sending commands to motor
+    // @Values: -1:Disabled,50:AUX1,51:AUX2,52:AUX3,53:AUX4,54:AUX5,55:AUX6
+    // @User: Standard
+    // @RebootRequired: True
+    AP_GROUPINFO("DE_PIN", 3, AP_Torqeedo, _pin_de, -1),
+
+    AP_GROUPEND
+};
+
+AP_Torqeedo::AP_Torqeedo()
+{
+    _singleton = this;
+    AP_Param::setup_object_defaults(this, var_info);
+}
+
+// initialise driver
+void AP_Torqeedo::init()
+{
+    // only init once
+    if (_initialised) {
+        return;
+    }
+
+    // create background thread to process serial input and output
+    if (!hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&AP_Torqeedo::loop, void), "torqeedo", 4096, AP_HAL::Scheduler::PRIORITY_RCOUT, 1)) {
```

Comment:
4kB stack?!  Yowsers!

Reply:
TBH, I have no idea what is a good number, again I just stole this from somewhere else.  For now I've reduced it to 2048.